src/Npgsql/NpgsqlTypes/NpgsqlRange.cs

```diff
@@ -19,161 +20,582 @@
+        /// <param name="y">The <see cref="NpgsqlRange{T}"/> on the right.</param>
+        /// <returns>
+        /// True if the <see cref="NpgsqlRange{T}"/> on the left is equal to the <see cref="NpgsqlRange{T}"/> on the right; otherwise, false.
+        /// </returns>
+        public static bool operator ==(NpgsqlRange<T> x, NpgsqlRange<T> y) => x.Equals(y);
+
+        /// <summary>
+        /// Indicates whether the <see cref="NpgsqlRange{T}"/> on the left is not equal to the <see cref="NpgsqlRange{T}"/> on the right.
+        /// </summary>
+        /// <param name="x">The <see cref="NpgsqlRange{T}"/> on the left.</param>
+        /// <param name="y">The <see cref="NpgsqlRange{T}"/> on the right.</param>
+        /// <returns>
+        /// True if the <see cref="NpgsqlRange{T}"/> on the left is not equal to the <see cref="NpgsqlRange{T}"/> on the right; otherwise, false.
+        /// </returns>
+        public static bool operator !=(NpgsqlRange<T> x, NpgsqlRange<T> y) => !x.Equals(y);
+
+        /// <inheritdoc />
+        public override bool Equals(object o) => o is NpgsqlRange<T> range && Equals(range);
+
+        /// <inheritdoc />
+        public bool Equals(NpgsqlRange<T> other)
         {
-            LowerBound = lowerBound;
-            UpperBound = upperBound;
-            Flags = flags;
-        }
+            if (Flags != other.Flags)
+                return false;
 
-        public static bool operator ==(NpgsqlRange<T> x, NpgsqlRange<T> y)
-            => x.IsEmpty == y.IsEmpty &&
-               (x.LowerBoundInfinite || y.LowerBoundInfinite || x.LowerBound.Equals(y.LowerBound)) &&
-               (x.UpperBoundInfinite || y.UpperBoundInfinite || x.UpperBound.Equals(y.UpperBound)) &&
-               x.LowerBoundIsInclusive == y.LowerBoundIsInclusive &&
-               x.UpperBoundIsInclusive == y.UpperBoundIsInclusive &&
-               x.LowerBoundInfinite == y.LowerBoundInfinite &&
-               x.UpperBoundInfinite == y.UpperBoundInfinite;
+            if (HasValueTypeBounds && HasEquatableBounds)
+                return
+                    ((IEquatable<T>)LowerBound).Equals(other.LowerBound) &&
+                    ((IEquatable<T>)UpperBound).Equals(other.UpperBound);
 
-        public static bool operator !=(NpgsqlRange<T> x, NpgsqlRange<T> y) => !(x == y);
+            if (HasValueTypeBounds)
+                return
+                    LowerBound.Equals(other.LowerBound) &&
+                    UpperBound.Equals(other.UpperBound);
 
-        public override bool Equals([CanBeNull] object o) => o is NpgsqlRange<T> && this == (NpgsqlRange<T>)o;
+            if (HasEquatableBounds)
+                return
+                    (((IEquatable<T>)LowerBound)?.Equals(other.LowerBound) ?? other.LowerBound == null) &&
+                    (((IEquatable<T>)UpperBound)?.Equals(other.UpperBound) ?? other.UpperBound == null);
 
-        public bool Equals(NpgsqlRange<T> other) => this == other;
+            return
```

Comment:
Distinguishing between value types and references types isn't really necessary here. You can keep the code that checks for null, and it should get elided by the JIT.